### PR TITLE
[FAU-414] Add support for campo keys

### PR DESCRIPTION
**Pull request by amiut**
**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines
- [ ] Tests for the changes have been added (for bug fixes/features)
- [ ] Docs have been added/updated (for bug fixes/features)

**What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
Feature

**What is the current behavior?** (You can also link to an open issue here)
https://inpsyde.atlassian.net/browse/FAU-414

**What is the new behavior (if this is a feature change)?**
Added support for campo keys

**Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)
No

**Related PRs:**
- https://github.com/RRZE-Webteam/FAU-Studium/pull/132
- https://github.com/RRZE-Webteam/FAU-Studium-Embed/pull/31

**Other information**:

### CODE REVIEW

File: src/Domain/CampoKeys.php
Comment (shvlv) [inline src/Domain/CampoKeys.php:1]: I really like Campo Keys, which is part of our domain model, as it should be.
What do you think if we remove unrequired getters and setters. I could imagine something like
```patch
Index: src/Infrastructure/Repository/CampoKeysRepository.php
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/Infrastructure/Repository/CampoKeysRepository.php b/src/Infrastructure/Repository/CampoKeysRepository.php
--- a/src/Infrastructure/Repository/CampoKeysRepository.php	(revision 4a72cbaf0318d8804397f2d94b6884a695bbd7af)
+++ b/src/Infrastructure/Repository/CampoKeysRepository.php	(date 1716362430281)
@@ -25,8 +25,6 @@
 
     public function degreeProgramCampoKeys(DegreeProgramId $degreeProgramId): CampoKeys
     {
-        $campoKeys = CampoKeys::empty();
-
         /** @var WP_Error|array<WP_Term> $terms */
         $terms = wp_get_post_terms(
             $degreeProgramId->asInt(),
@@ -34,8 +32,10 @@
         );
 
         if ($terms instanceof WP_Error) {
-            return $campoKeys;
+            return CampoKeys::empty();
         }
+
+        $map = [];
 
         foreach ($terms as $term) {
             $campoKey = (string) get_term_meta($term->term_id, self::CAMPOKEY_TERM_META_KEY, true);
@@ -50,10 +50,10 @@
                 continue;
             }
 
-            $campoKeys->set($campoKeyType, $campoKey);
+            $map[$campoKeyType] = $campoKey;
         }
 
-        return $campoKeys;
+        return CampoKeys::fromArray($map);
     }
 
     /**
Index: src/Domain/CampoKeys.php
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/Domain/CampoKeys.php b/src/Domain/CampoKeys.php
--- a/src/Domain/CampoKeys.php	(revision 4a72cbaf0318d8804397f2d94b6884a695bbd7af)
+++ b/src/Domain/CampoKeys.php	(date 1716362430285)
@@ -4,8 +4,9 @@
 
 namespace Fau\DegreeProgram\Common\Domain;
 
-use InvalidArgumentException;
-
+/**
+ * @psalm-type CampoKeysMap = array<value-of<self::SUPPORTED_CAMPO_KEYS>, string>
+ */
 final class CampoKeys
 {
     public const SCHEMA = [
@@ -46,85 +47,50 @@
 
     private const HIS_CODE_DELIMITER = '|';
 
-    /** @var array<string, string> */
-    private array $map = [];
-
-    private function __construct()
-    {
+    private function __construct(
+        /**
+         * @var CampoKeysMap $map
+         */
+        private array $map
+    ) {
     }
 
     public static function empty(): self
     {
-        return new self();
+        return new self([]);
     }
 
     /**
-     * @psalm-param array<string, string> $data
+     * @param CampoKeysMap $map
      */
-    public static function fromArray(array $data): self
+    public static function fromArray(array $map): self
     {
-        $instance = new self();
-
-        foreach ($data as $key => $value) {
-            $instance->set($key, $value);
-        }
-
-        return $instance;
+        return new self($map);
     }
 
     public static function fromHisCode(string $hisCode): self
     {
         $parts = explode(self::HIS_CODE_DELIMITER, $hisCode);
+        $map = [];
 
-        $instance = new self();
 
         if (isset($parts[0])) {
-            $instance->set(DegreeProgram::DEGREE, $parts[0]);
+            $map[DegreeProgram::DEGREE] = $parts[0];
         }
 
         if (isset($parts[1])) {
-            $instance->set(DegreeProgram::AREA_OF_STUDY, $parts[1]);
+            $map[DegreeProgram::AREA_OF_STUDY] = $parts[1];
         }
 
         if (isset($parts[6])) {
-            $instance->set(DegreeProgram::LOCATION, $parts[6]);
+            $map[DegreeProgram::LOCATION] = $parts[6];
         }
-
-        return $instance;
-    }
-
-    public function set(string $key, string $value): self
-    {
-        if (! in_array($key, self::SUPPORTED_CAMPO_KEYS, true)) {
-            throw new InvalidArgumentException('Unsupported field key.');
-        }
-
-        $this->map[$key] = $value;
-        return $this;
-    }
-
-    public function degree(): ?string
-    {
-        return $this->get(DegreeProgram::DEGREE);
-    }
-
-    public function areaOfStudy(): ?string
-    {
-        return $this->get(DegreeProgram::AREA_OF_STUDY);
-    }
 
-    public function studyLocation(): ?string
-    {
-        return $this->get(DegreeProgram::LOCATION);
-    }
-
-    public function get(string $key): ?string
-    {
-        return $this->map[$key] ?? null;
+        return new self($map);
     }
 
     /**
-     * @return array<string, string>
+     * @return CampoKeysMap
      */
     public function asArray(): array
     {

```

File: src/Domain/CampoKeys.php [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+                'type' => 'string',
+            ],
+            DegreeProgram::AREA_OF_STUDY => [
+                'type' => 'string',
+            ],
+            DegreeProgram::LOCATION => [
+                'type' => 'string',
+            ],
+        ],
+    ];
```
Comment (shvlv) [inline src/Domain/CampoKeys.php:40]: Is it intentional that the required schema is identical to the draft one? Should we require those properties, or should we not require Campo key parts for terms?

File: src/Infrastructure/Repository/CampoKeysRepository.php [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+use Fau\DegreeProgram\Common\Domain\DegreeProgramId;
+use Fau\DegreeProgram\Common\Infrastructure\Content\Taxonomy\AreaOfStudyTaxonomy;
+use Fau\DegreeProgram\Common\Infrastructure\Content\Taxonomy\DegreeTaxonomy;
+use Fau\DegreeProgram\Common\Infrastructure\Content\Taxonomy\StudyLocationTaxonomy;
+use WP_Error;
+use WP_Term;
+
+final class CampoKeysRepository
+{
+    public const CAMPO_KEYS_TOTAXONOMY_MAP = [
```
Comment (shvlv) [inline src/Infrastructure/Repository/CampoKeysRepository.php]:
```suggestion
    public const TAXONOMY_TO_CAMPO_KEY_MAP = [
```

File: src/Infrastructure/Repository/CampoKeysRepository.php [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+        DegreeTaxonomy::KEY => DegreeProgram::DEGREE,
+        StudyLocationTaxonomy::KEY => DegreeProgram::LOCATION,
+        AreaOfStudyTaxonomy::KEY => DegreeProgram::AREA_OF_STUDY,
+    ];
+
+    public const CAMPOKEY_TERM_META_KEY = 'uniquename';
```
Comment (shvlv) [inline src/Infrastructure/Repository/CampoKeysRepository.php]:
```suggestion
    public const CAMPO_KEY_TERM_META_KEY = 'uniquename';
```

File: src/Infrastructure/Repository/CampoKeysRepository.php [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+
+        foreach (self::CAMPO_KEYS_TOTAXONOMY_MAP as $taxonomy => $campoKeyType) {
+            $campoKey = $campoKeys[$campoKeyType] ?? '';
+
+            if ($campoKey === '') {
+                continue;
+            }
+
+            $term = $this->findTermByCampoKey($taxonomy, $campoKey);
+            $result[$taxonomy] = ! is_null($term) ? $term->term_id : 0;
```
Comment (shvlv) [inline src/Infrastructure/Repository/CampoKeysRepository.php]: I would not add an empty value to the result.

**Review by shvlv — APPROVED** [review]
Thanks! Looks good to me in general. Left minor comments.

**Review by zhyian — APPROVED** [review]
LGTM and it's functioning well on my local setup. I just have two small issues, but I think the reason might be my cached data, which doesn't include campo keys yet.

File: src/Domain/CampoKeys.php (hunk 1 above)
Comment (amiut) [reply src/Domain/CampoKeys.php:40]: Since entering campokeys for terms is optional, I used the same schema for required as well, otherwise it can prevent saving the degree program if one of the terms doesn't have a campo key.

File: src/Infrastructure/Repository/CampoKeysRepository.php (hunk 4 above)
Comment (amiut) [reply src/Infrastructure/Repository/CampoKeysRepository.php]: the problem is if we don't pass `0` term id it will continue finding results with wrong HIS codes.

Comment (shvlv) [reply src/Infrastructure/Repository/CampoKeysRepository.php]: I meant something like
```php
if ($term instanceof \WP_Term) {
   $result[$taxonomy] = $term->term_id;
}
```

Does it work?

Comment (amiut) [reply src/Infrastructure/Repository/CampoKeysRepository.php]: the problem still remains the same, consider a scenario where editor passes an HIS code like `185|80`
but there is no degree with campo key `185` (a.k.a the `$term` here is null) and as a result the tax query item for degree isn't built [here](https://github.com/RRZE-Webteam/FAU-Studium-Common/blob/0816b3af8d637e943e0d8cd1ea1f80349f19a79e/src/Infrastructure/Repository/WpQueryArgsBuilder.php#L103) so the tax query doesn't include degree at all and will lead to wrong results, while it shouldn't show anything if a part of HIS code is wrong.

Comment (shvlv) [reply src/Infrastructure/Repository/CampoKeysRepository.php]: Oh, okay. So, we create an invalid WP_Query to enforce empty results. We could throw and catch at a higher level the exception instead. On the other side, we don't validate other filters/terms IDs... Could you add a comment to explain this?

Comment (amiut) [reply src/Infrastructure/Repository/CampoKeysRepository.php]: sure that makes more sense, I followed the exception approach with a clarifying comment.

File: src/Domain/CampoKeys.php [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+
+        if (isset($parts[1])) {
+            $map[DegreeProgram::AREA_OF_STUDY] = $parts[1];
+        }
+
+        if (isset($parts[6])) {
+            $map[DegreeProgram::LOCATION] = $parts[6];
+        }
+
+        return new self($map);
```
Comment (tyrann0us) [inline src/Domain/CampoKeys.php]: Given that we might support more campo key tuples in the future, does it make sense to reduce verbosity by something like this?
```suggestion
        $map = [
            DegreeProgram::DEGREE => $parts[0] ?? null,
            DegreeProgram::AREA_OF_STUDY => $parts[1] ?? null,
            DegreeProgram::LOCATION => $parts[6] ?? null,
        ];

        return new self(array_filter($map, fn($value) => !is_null($value)));
```

File: src/Infrastructure/Repository/WpQueryArgsBuilder.php [hunk 6]
```diff
@@ -84,9 +88,47 @@ public function build(CollectionCriteria $criteria): WpQueryArgs
...
+                        $termId,
+                    ],
+                ];
+            }
+
+            return $queryArgs->withTaxQueryItem($taxQueryItem);
+        } catch (RuntimeException) {
+            /*
+             * Return an empty result if one or more campo keys in HIS code are not matched to any terms.
+             * Otherwise invalid HIS codes would be matched to false results.
```
Comment (tyrann0us) [inline src/Infrastructure/Repository/WpQueryArgsBuilder.php:126]: Please ignore if you want to keep the name "HIS code".
```suggestion
             * Return an empty result if one or more Campo Keys can't be matched to any terms.
             * Otherwise, invalid Campo Keys will be matched to false results.
```

**Review by o-samaras — APPROVED** [review]
LGTM! Thank you

**Review by tyrann0us — COMMENTED** [review]
Thanks for working on this! Left a couple of minor comments. Also, do I understand correctly that naming-wise, you consider an HIS code to consist of multiple Campo Keys? So Campo Keys are, e.g., "08", "185", or "AB1", and an HIS code would be, e.g., "**08**|**185**|-|-|H|**AB1**|E|P|V|1|"? I don't know if this matches the naming in the client's business domain. The "Campo Key" term meta field name points in this direction, but it could also be like this to avoid confusing users with other names (e.g., "Campo Key Tuple").

Long story short, let's keep everything as-is, but in the future, we could rename everything from "HIS code", `hisCode`, or `HIS_CODE` to "Campo Key",  `campoKey`, or `CAMPO_KEY` to prevent confusion. An outside developer probably doesn't know that "HIS code" and "Campo Key" is related on first glance.

**Comment by amiut** [comment]
> Thanks for working on this! Left a couple of minor comments. Also, do I understand correctly that naming-wise, you consider an HIS code to consist of multiple Campo Keys? So Campo Keys are, e.g., "08", "185", or "AB1", and an HIS code would be, e.g., "**08**|**185**|-|-|H|**AB1**|E|P|V|1|"? I don't know if this matches the naming in the client's business domain. The "Campo Key" term meta field name points in this direction, but it could also be like this to avoid confusing users with other names (e.g., "Campo Key Tuple").
> 
> Long story short, let's keep everything as-is, but in the future, we could rename everything from "HIS code", `hisCode`, or `HIS_CODE` to "Campo Key", `campoKey`, or `CAMPO_KEY` to prevent confusion. An outside developer probably doesn't know that "HIS code" and "Campo Key" is related on first glance.

I think there correct is still not clear for example the title of this ticket https://inpsyde.atlassian.net/browse/FAU-401 calls it **"HIS-identifier"**, what should we call the individual members in term meta if "Campo keys" can be considered the whole `|`-separated string?

**Review by tyrann0us — COMMENTED** [review]
> what should we call the individual members in term meta if "Campo keys" can be considered the whole |-separated string?

During the requirement engineering phase, we called it "Campo Key tuples" or just "tuples".
